Panic at tree service

neofs-node 0.28.1-112-g4a50bdfa (`tree-service` branch).

During launch of [s3-tests](), neofs-node throws a panic and restarts at [versioned_concurrent_object_create_concurrent_remove](https://github.com/nspcc-dev/s3-tests/blob/5b19cc1bf61c0f96231dee787fde07b2f91d6368/s3tests_boto3/functional/test_s3.py#L9178) test.

```
2022-06-06T07:33:59.495Z        info    log/log.go:13   local object storage operation  {"address": "H4RR2jRsaaEZZfgYEgiFn23sUn8fLX2vyqkmA5HpWJzS/Edr8ZjWmMRZmrLFhU4NDQ1XSxU4f35LYuW4sBLT3FBaH", "op": "metabase PUT"}
2022-06-06T07:33:59.495Z        debug   delete/delete.go:37     operation finished successfully {"component": "Object.Delete service", "request": "DELETE", "address": "H4RR2jRsaaEZZfgYEgiFn23sUn8fLX2vyqkmA5HpWJzS/6RzNFKpmVr7EkgiRMS7xrZrfQNaZBEmJjVLfj5vAmcN8", "local": false, "with session": true, "with bearer": true}
panic: runtime error: index out of range [7] with length 0                                                                                                                                                            
                                                                                                                                                                                                                      
goroutine 72999 [running]:                                                                                                                                                                                            
encoding/binary.littleEndian.Uint64(...)                                                                                                                                                                              
        encoding/binary/binary.go:77                                                                                                                                                                                  
github.com/nspcc-dev/neofs-node/pkg/local_object_storage/pilorama.(*boltForest).TreeGetByPath.func1(0xc0042d4b60)                                                                                                     
        github.com/nspcc-dev/neofs-node/pkg/local_object_storage/pilorama/boltdb.go:451 +0x5fe                                                                                                                        
go.etcd.io/bbolt.(*DB).View(0xc0034f26f0?, 0xc0012d7670)                                                                                                                                                              
        go.etcd.io/bbolt@v1.3.6/db.go:772 +0x82                                                                                                                                                                       
github.com/nspcc-dev/neofs-node/pkg/local_object_storage/pilorama.(*boltForest).TreeGetByPath(0xc003be03d0?, {0xee, 0x9b, 0x56, 0xae, 0x24, 0xef, 0xde, 0x42, 0x26, ...}, ...)                                        
        github.com/nspcc-dev/neofs-node/pkg/local_object_storage/pilorama/boltdb.go:422 +0x12c                                                                                                                        
github.com/nspcc-dev/neofs-node/pkg/local_object_storage/shard.(*Shard).TreeGetByPath(...)                                                                                                                            
        github.com/nspcc-dev/neofs-node/pkg/local_object_storage/shard/tree.go:27                                                                                                                                     
github.com/nspcc-dev/neofs-node/pkg/local_object_storage/engine.(*StorageEngine).TreeGetByPath(0xc0003d2500, {0xee, 0x9b, 0x56, 0xae, 0x24, 0xef, 0xde, 0x42, 0x26, ...}, ...)                                        
        github.com/nspcc-dev/neofs-node/pkg/local_object_storage/engine/tree.go:70 +0x17e                                                                                                                             
github.com/nspcc-dev/neofs-node/pkg/services/tree.(*Service).GetNodeByPath(0xc0001d5130, {0x11461e8, 0xc003b201e0}, 0xc0027f4140)                                                                                     
        github.com/nspcc-dev/neofs-node/pkg/services/tree/service.go:303 +0x2f2                                                                                                                                       
github.com/nspcc-dev/neofs-node/pkg/services/tree._TreeService_GetNodeByPath_Handler({0xf8afa0?, 0xc0001d5130}, {0x11461e8, 0xc003b201e0}, 0xc003ac6d80, 0x0)                                                         
        github.com/nspcc-dev/neofs-node/pkg/services/tree/service_grpc.pb.go:309 +0x170                                                                                                                               
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0004aa000, {0x114b2a0, 0xc000188600}, 0xc00166b200, 0xc00029bc20, 0x1867ee0, 0x0)                                                                                 
        google.golang.org/grpc@v1.41.0/server.go:1279 +0xccf                                                                                                                                                          
google.golang.org/grpc.(*Server).handleStream(0xc0004aa000, {0x114b2a0, 0xc000188600}, 0xc00166b200, 0x0)                                                                                                             
        google.golang.org/grpc@v1.41.0/server.go:1608 +0xa1b                                                                                                                                                          
google.golang.org/grpc.(*Server).serveStreams.func1.2()                                                                                                                                                               
        google.golang.org/grpc@v1.41.0/server.go:923 +0x98                                                                                                                                                            
created by google.golang.org/grpc.(*Server).serveStreams.func1                                                                                                                                                        
        google.golang.org/grpc@v1.41.0/server.go:921 +0x28a                                                                                                                                                           
```

I can go further to see what requests S3 Gateway produces, if needed.

--- a/pkg/local_object_storage/pilorama/boltdb.go
+++ b/pkg/local_object_storage/pilorama/boltdb.go
@@ -270,9 +270,7 @@ func (t *boltForest) applyOperation(logBucket, treeBucket *bbolt.Bucket, lm *Log
 }
 
 func (t *boltForest) do(lb *bbolt.Bucket, b *bbolt.Bucket, key []byte, op *LogMove) error {
-	shouldPut := !t.isAncestor(b, key, op.Child, op.Parent) &&
-		!(op.Parent != 0 && op.Parent != TrashID && b.Get(timestampKey(key, op.Parent)) == nil)
-	shouldRemove := op.Parent == TrashID
+	shouldPut := !t.isAncestor(b, key, op.Child, op.Parent)
 
 	currParent := b.Get(parentKey(key, op.Child))
 	if currParent != nil { // node is already in tree
@@ -290,16 +288,6 @@ func (t *boltForest) do(lb *bbolt.Bucket, b *bbolt.Bucket, key []byte, op *LogMo
 
 	if !shouldPut {
 		return nil
-	}
-
-	if shouldRemove {
-		if currParent != nil {
-			p := binary.LittleEndian.Uint64(currParent)
-			if err := b.Delete(childrenKey(key, op.Child, p)); err != nil {
-				return err
-			}
-		}
-		return t.removeNode(b, key, op.Child, op.Parent)
 	}
 
 	if currParent == nil {

--- a/pkg/local_object_storage/pilorama/forest_test.go
+++ b/pkg/local_object_storage/pilorama/forest_test.go
@@ -354,16 +354,17 @@ func testForestTreeApply(t *testing.T, constructor func(t testing.TB) Forest) {
 		testMeta(t, s, cid, treeID, 11, 10, meta)
 
 		testApply(t, s, 10, TrashID, Meta{Time: 2, Items: []KeyValue{{"parent", []byte{2}}}})
-		testMeta(t, s, cid, treeID, 11, 0, Meta{})
+		testMeta(t, s, cid, treeID, 11, 10, meta)
 	})
 	t.Run("add a child to non-existent parent, then add a parent", func(t *testing.T) {
 		s := constructor(t)
 
-		testApply(t, s, 11, 10, Meta{Time: 1, Items: []KeyValue{{"child", []byte{3}}}})
-		testMeta(t, s, cid, treeID, 11, 0, Meta{})
+		meta := Meta{Time: 1, Items: []KeyValue{{"child", []byte{3}}}}
+		testApply(t, s, 11, 10, meta)
+		testMeta(t, s, cid, treeID, 11, 10, meta)
 
 		testApply(t, s, 10, 0, Meta{Time: 2, Items: []KeyValue{{"grand", []byte{1}}}})
-		testMeta(t, s, cid, treeID, 11, 0, Meta{})
+		testMeta(t, s, cid, treeID, 11, 10, meta)
 	})
 }
 

--- a/pkg/local_object_storage/pilorama/inmemory.go
+++ b/pkg/local_object_storage/pilorama/inmemory.go
@@ -86,11 +86,7 @@ func (s *state) do(op *Move) LogMove {
 		},
 	}
 
-	_, parentInTree := s.tree.infoMap[op.Parent]
-	shouldPut := !s.tree.isAncestor(op.Child, op.Parent) &&
-		!(op.Parent != 0 && op.Parent != TrashID && !parentInTree)
-	shouldRemove := op.Parent == TrashID
-
+	shouldPut := !s.tree.isAncestor(op.Child, op.Parent)
 	p, ok := s.tree.infoMap[op.Child]
 	if ok {
 		lm.HasOld = true
@@ -98,14 +94,6 @@ func (s *state) do(op *Move) LogMove {
 	}
 
 	if !shouldPut {
-		return lm
-	}
-
-	if shouldRemove {
-		if ok {
-			s.removeChild(op.Child, p.Parent)
-		}
-		delete(s.tree.infoMap, op.Child)
 		return lm
 	}
 

